GoogleForCreators/web-stories-wp
# Taxonomy: freeform interaction and base infrastructure

## Context
Part of initial taxonomy FE infrastructure and WP API interactions.



## Summary
Adds: 
- Base Taxonomy provider setup
- Hydrating initial taxonomy from loaded story
- Updating taxonomy in story state
- Saving Taxonomy when saving story
- Creating feeform taxonomy terms
- Tying together freeform taxonomy input with taxonomy BE interactions



## Relevant Technical Choices

**Shape of State**
Basic idea here is that we store all taxonomies in a nested dictionary/cache. Shape should follow:
```
TermCache {
  [Taxonomy['rest_base']]: {
    [Term['slug]]: Term
  }
}
```

We keep track of selected/post associated terms in the following shape:
```
SelectedSlugs {
  [Taxonomy['rest_base]]: Array<Term['slug']>
}
```

**Initial Hydration**
Terms are pulled off the story's `_embed['wp:term']` property on initial load and used to populate the initial cache as well as the initial selected slugs.

**User Interaction**
Terms are fetched from the BE and populate the cache from either:
  - Debounced callback to freeform input requesting `/<taxonomy>?search=<current_input_val>`
  - Post to create tag on submission of freeform input

This allows us to update selected slugs in real time without waiting for the response from the BE, and we sync up the story's selected term IDs every time either the Selected Slugs or Term Cache update.



## To-do
NA

Gonna increase test coverage here. karma and provider unit:
https://app.zenhub.com/workspaces/web-stories-5e94d3aced449034e1e9b226/issues/google/web-stories-wp/8842



## User-facing changes
NA


## Testing Instructions
- Upload this plugin and activate it to register a couple taxonomies onto the story post type:
[web-stories-taxonomy-example.php.zip](https://github.com/google/web-stories-wp/files/7164177/web-stories-taxonomy-example.php.zip)

- Make sure the taxonomy flag is enabled.
- Go to editor & open up the document panel
- scroll down to tags & categories panel
- add some taxonomies
- save story
- refresh page
- see that taxonomies are persisted





This PR can be tested by following these steps:

1.


## Reviews

### Does this PR have a security-related impact?



### Does this PR change what data or activity we track or use?



### Does this PR have a legal-related impact?



## Checklist




---



Fixes #8838 
Fixes #8839
